### PR TITLE
Configuration API

### PR DESCRIPTION
**Pull request by gkillough**
The framework is laid an everything is refactored and seems to be working correctly (based on smoke tests). More refactoring work will be done in a separate branch to minimize the scope of each branch.

### CODE REVIEW

**Comment by ghost** [comment]
## Black Duck Security Report

Merging #69 into master will not change security risk.

[Click here to see full report](https://copilot.blackducksoftware.com/github/repos/blackducksoftware/blackduck-jira/pull-requests/69)

File: src/main/java/com/blackducksoftware/integration/jira/common/settings/PluginSettingsWrapper.java [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+//
+//import java.util.Optional;
+//
+//import org.apache.commons.lang3.StringUtils;
+//import org.apache.commons.lang3.math.NumberUtils;
+//
+//import com.atlassian.sal.api.pluginsettings.PluginSettings;
+//
+//// TODO change these methods to all return optionals.
+//public class PluginSettingsWrapper {
```
Comment (jamesrichard91) [inline src/main/java/com/blackducksoftware/integration/jira/common/settings/PluginSettingsWrapper.java]: Should this file just be deleted?

Comment (gkillough) [reply src/main/java/com/blackducksoftware/integration/jira/common/settings/PluginSettingsWrapper.java]: Yes, I meant to do this.

File: src/main/java/com/blackducksoftware/integration/jira/config/PluginConfigurationDetails.java [hunk 2]
```diff
@@ -21,172 +21,195 @@
...
+//package com.blackducksoftware.integration.jira.config;
+//
+//import org.apache.commons.lang3.math.NumberUtils;
+//
+//import com.atlassian.sal.api.pluginsettings.PluginSettings;
+//import com.blackducksoftware.integration.jira.common.settings.PluginSettingsWrapper;
+//import com.synopsys.integration.blackduck.configuration.BlackDuckServerConfig;
+//import com.synopsys.integration.blackduck.configuration.BlackDuckServerConfigBuilder;
+//
+//public class PluginConfigurationDetails {
```
Comment (jamesrichard91) [inline src/main/java/com/blackducksoftware/integration/jira/config/PluginConfigurationDetails.java]: Should we just delete this file?

**Review by jamesrichard91 — APPROVED** [review]

File: src/main/java/com/blackducksoftware/integration/jira/common/settings/JiraSettingsAccessor.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    public PluginErrorAccessor createPluginErrorAccessor() {
+        return new PluginErrorAccessor(this);
+    }
+
+    public GlobalConfigurationAccessor createGlobalConfigurationAccessor() {
+        return new GlobalConfigurationAccessor(this);
+    }
+
+    public Object getObjectValue(final String key) {
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/common/settings/JiraSettingsAccessor.java:52]: Should we have all these getters return optionals?

File: src/main/java/com/blackducksoftware/integration/jira/common/settings/model/PluginGroupsConfigModel.java [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.synopsys.integration.util.Stringable;
+
+public class PluginGroupsConfigModel extends Stringable {
+    public static final String BLACK_DUCK_GROUPS_LIST_DELIMETER = ",";
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/common/settings/model/PluginGroupsConfigModel.java:35]: Is it possible for a BD group to contain a comma? If so, we could have bugs regarding this in the future.

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 5]
```diff
@@ -52,26 +59,30 @@
...
-        this.projectMappingConfigActions = new ProjectMappingConfigActions(pluginSettingsFactory, workflowHelper);
+        this.projectMappingConfigActions = new ProjectMappingConfigActions(jiraSettingsAccessor, workflowHelper);
     }
 
     public BlackDuckJiraConfigSerializable getCreator() {
         final BlackDuckJiraConfigSerializable txConfig = new BlackDuckJiraConfigSerializable();
-        final String creator = pluginSettingsWrapper.getIssueCreatorUser();
+
+        final PluginIssueCreationConfigModel issueCreationConfig = globalConfigurationAccessor.getIssueCreationConfig();
+        final String creator = issueCreationConfig.getGeneral().getDefaultIssueCreator();
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:85]: you could add some helper methods in GlobalConfigurationAccessor that calls these methods directly so you don't have to have as many dots.

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 6]
```diff
@@ -105,7 +116,8 @@ public BlackDuckJiraConfigSerializable getJiraProjects() {
     public BlackDuckJiraConfigSerializable getCommentOnIssueUpdates() {
         logger.debug("GET /comment/updatechoice transaction");
         final BlackDuckJiraConfigSerializable txConfig = new BlackDuckJiraConfigSerializable();
-        final Boolean choice = pluginSettingsWrapper.getCommentOnIssuesUpdatesChoice();
+        final PluginIssueCreationConfigModel issueCreationConfig = globalConfigurationAccessor.getIssueCreationConfig();
+        final Boolean choice = issueCreationConfig.getTicketCriteria().getCommentOnIssueUpdates();
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:120]: Same here

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 7]
```diff
@@ -114,15 +126,19 @@ public BlackDuckJiraConfigSerializable getCommentOnIssueUpdates() {
     public BlackDuckJiraConfigSerializable getProjectReviewerNotificationsChoice() {
         logger.debug("GET /project/reviewerChoice transaction");
         final BlackDuckJiraConfigSerializable txConfig = new BlackDuckJiraConfigSerializable();
-        final Boolean choice = pluginSettingsWrapper.getProjectReviewerNotificationsChoice();
+        final PluginIssueCreationConfigModel issueCreationConfig = globalConfigurationAccessor.getIssueCreationConfig();
+        final Boolean choice = issueCreationConfig.getTicketCriteria().getAddComponentReviewerToTickets();
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:130]: Same here

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 8]
```diff
@@ -114,15 +126,19 @@ public BlackDuckJiraConfigSerializable getCommentOnIssueUpdates() {
...
         return txConfig;
     }
 
     public BlackDuckJiraConfigSerializable getInterval() {
         final BlackDuckJiraConfigSerializable txConfig = new BlackDuckJiraConfigSerializable();
-        pluginSettingsWrapper.getIntervalBetweenChecks().ifPresent(integer -> txConfig.setIntervalBetweenChecks(String.valueOf(integer)));
+        final PluginIssueCreationConfigModel issueCreationConfig = globalConfigurationAccessor.getIssueCreationConfig();
+
+        final Optional<Integer> intervalBetweenChecks = issueCreationConfig.getGeneral().getInterval();
+        intervalBetweenChecks.map(String::valueOf).ifPresent(txConfig::setIntervalBetweenChecks);
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:141]: same here

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 9]
```diff
@@ -134,28 +150,36 @@ public BlackDuckJiraConfigSerializable updateConfig(final BlackDuckJiraConfigSer
...
+
         final String issueCreatorJiraUser = config.getCreator();
-        logger.debug("Setting issue creator jira user to: " + issueCreatorJiraUser);
-        pluginSettingsWrapper.setIssueCreatorUser(issueCreatorJiraUser);
-        pluginSettingsWrapper.setPolicyRulesJson(config.getPolicyRulesJson());
-        pluginSettingsWrapper.setProjectMappingsJson(config.getHubProjectMappingsJson());
-        final String username = authorizationChecker.getUsername(request).orElse(null);
-        pluginSettingsWrapper.setJiraAdminUser(username);
-        final Optional<Integer> previousInterval = pluginSettingsWrapper.getIntervalBetweenChecks();
+        final Optional<Integer> previousInterval = previousIssueCreationConfig.getGeneral().getInterval();
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:165]: previousIssueCreationConfig doesn't look like it's used anywhere else and can probably be collapsed.

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 10]
```diff
@@ -164,7 +188,9 @@ private void validateCreator(final BlackDuckJiraConfigSerializable config) {
         if (StringUtils.isBlank(config.getCreator())) {
             config.setGeneralSettingsError(JiraConfigErrorStrings.NO_CREATOR_SPECIFIED_ERROR);
         }
-        if (authorizationChecker.isValidAuthorization(config.getCreator(), pluginSettingsWrapper.getParsedBlackDuckConfigGroups())) {
+
+        final PluginGroupsConfigModel groupsConfig = globalConfigurationAccessor.getGroupsConfig();
+        if (authorizationChecker.isValidAuthorization(config.getCreator(), groupsConfig.getGroups())) {
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:193]: groups can be collapsed into a single method

File: src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java [hunk 11]
```diff
@@ -173,7 +199,8 @@ private void validateCreator(final BlackDuckJiraConfigSerializable config) {
 
     private SortedSet<String> getIssueCreatorCandidates() {
         final SortedSet<String> jiraUsernames = new TreeSet<>();
-        final String[] groupNames = pluginSettingsWrapper.getParsedBlackDuckConfigGroups();
+        final PluginGroupsConfigModel groupsConfig = globalConfigurationAccessor.getGroupsConfig();
+        final Collection<String> groupNames = groupsConfig.getGroups();
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/config/controller/action/IssueCreationConfigActions.java:203]: same groups call

File: src/main/java/com/blackducksoftware/integration/jira/task/BlackDuckMonitor.java [hunk 12]
```diff
@@ -158,11 +159,21 @@ private long getIntervalMillisec() {
...
             return DEFAULT_INTERVAL_MILLISEC;
         }
-        final String intervalString = (String) pluginSettings.get(PluginConfigKeys.BLACKDUCK_CONFIG_JIRA_INTERVAL_BETWEEN_CHECKS);
-        if (intervalString == null) {
+
+        final JiraSettingsAccessor jiraSettingsAccessor = new JiraSettingsAccessor(pluginSettings);
+        final GlobalConfigurationAccessor globalConfigurationAccessor = new GlobalConfigurationAccessor(jiraSettingsAccessor);
+        final PluginIssueCreationConfigModel issueCreationConfig = globalConfigurationAccessor.getIssueCreationConfig();
+        final GeneralIssueCreationConfigModel generalConfig = issueCreationConfig.getGeneral();
+        final Optional<String> optionalIntervalString = generalConfig.getInterval().map(String::valueOf);
```
Comment (bamandel) [inline src/main/java/com/blackducksoftware/integration/jira/task/BlackDuckMonitor.java:167]: Another example

**Review by bamandel — COMMENTED** [review]
Most of the comments all regard the same idea. If you don't like the idea, then everything looks fine.

**Review by psantos1113 — APPROVED** [review]

**Pull request merged**